### PR TITLE
actualize Vagrant and fix problems on orchestrator-agent api with sqlite backend

### PR DESCRIPTION
**Pull request by Slach**
Related issue: https://github.com/github/orchestrator/issues/438

### Description

This PR actualize Vagrant and fix problems on orchestrator-agent api with sqlite backend
i hope this help use innodbbackupex for data seeding between two mysql nodes

> In case this PR introduced Go code changes:

- [x] contributed code is using same conventions as original code
- [x] code is formatted via `gofmt` (please avoid `goimports`)
- [x] code is built via `./build.sh`
- [x] code is tested via `go test ./go/...`

### CODE REVIEW

File: go/agent/agent_dao.go [hunk 1]
```diff
@@ -547,8 +547,8 @@ func AbortSeed(seedId int64) error {
 }
 
 // PostCopy will request an agent to invoke post-copy commands
-func PostCopy(hostname string) (Agent, error) {
-	return executeAgentCommand(hostname, "post-copy", nil)
+func PostCopy(hostname, sourceHostname string) (Agent, error) {
```
Comment (shlomi-noach) [inline go/agent/agent_dao.go:550]: Is this change going to break behavior for existing users?

Comment (Slach) [reply go/agent/agent_dao.go:550]: could you help me figure out howto add optional parameter to martini routes?

Comment (shlomi-noach) [reply go/agent/agent_dao.go:550]: Sure. Consider, for example,

https://github.com/github/orchestrator/blob/f465c53809370697566ca84b418fd43f7fe41ca9/go/http/api.go#L3291-L3292

and then

https://github.com/github/orchestrator/blob/f465c53809370697566ca84b418fd43f7fe41ca9/go/http/api.go#L2696-L2726

Comment (Slach) [reply go/agent/agent_dao.go:550]: @shlomi-noach ok. 
i add backwards compatibility on orchestrator-agent here
https://github.com/github/orchestrator-agent/pull/20/files#diff-fbdc8cde28a71961e53c2e83e2483c9aL619

and add some other improvments and tested it with percona-xtrabackup
currenly in my setup i have successfull data seed between nodes over netcat + xtrabackup + xbstream

File: etc/init.d/orchestrator.bash [hunk 2]
```diff
@@ -54,8 +54,9 @@ start_daemon () {
 	post_start_daemon_hook 1>&2
 }
 
-# The file /etc/orchestrator_profile can be used to inject pre-service execution
+# This files can be used to inject pre-service execution
 # scripts, such as exporting variables or whatever. It's yours!
+[ -f /etc/default/orchestrator ] && . /etc/default/orchestrator
```
Comment (shlomi-noach) [inline etc/init.d/orchestrator.bash:59]: Please explain why `/etc/default/orchestrator`? IF anything, I'd add `/etc/profile.d/orchestrator` which is what I should have done in the first place.

**Comment by shlomi-noach** [comment]
Thank you! This PR is growing bigger and deals with different problems, so it's becoming difficult for me to review it as a whole. Can you please elaborate on "i add backwards compatibility on orchestrator-agent here" ? Where is the backwards compatibility found?

File: etc/init.d/orchestrator.bash (hunk 2 above)
Comment (Slach) [reply etc/init.d/orchestrator.bash:59]: /etc/default it's a default behavion on debian based distributives

File: go/agent/agent_dao.go [hunk 3]
```diff
@@ -164,7 +164,7 @@ func ReadOutdatedAgentsHosts() ([]string, error) {
 		from
 			host_agent
 		where
-			IFNULL(last_checked < now() - interval ? minute, true)
+			IFNULL(last_checked < now() - interval ? minute, 1)
```
Comment (Slach) [inline go/agent/agent_dao.go:167]: this change need for SQLlite compatibility cause SQLite desn't have BOOLEAN literal

File: go/agent/agent_dao.go [hunk 4]
```diff
@@ -547,8 +547,8 @@ func AbortSeed(seedId int64) error {
 }
 
 // PostCopy will request an agent to invoke post-copy commands
-func PostCopy(hostname string) (Agent, error) {
-	return executeAgentCommand(hostname, "post-copy", nil)
+func PostCopy(hostname, sourceHostname string) (Agent, error) {
+	return executeAgentCommand(hostname, fmt.Sprintf("post-copy/?sourceHost=%s", sourceHostname), nil)
```
Comment (Slach) [inline go/agent/agent_dao.go:551]: there is backwards compatibility

**Comment by Slach** [comment]
>Can you please elaborate on "i add backwards compatibility on orchestrator-agent here" ?
backwards behavior it's here

https://github.com/github/orchestrator-agent/pull/20/files#diff-fbdc8cde28a71961e53c2e83e2483c9aL619
and here
https://github.com/github/orchestrator/pull/445/files#diff-17473f4f3d0788c153babef6ddbba1faR551

File: etc/init.d/orchestrator.bash (hunk 2 above)
Comment (Slach) [reply etc/init.d/orchestrator.bash:59]: ok, i add usage  ```/etc/profile.d/orchestrator``` and ```/etc/default/orchestrator``` 

```/etc/profile.d/orchestrator``` will have major priority

**Comment by shlomi-noach** [comment]
I will get back on this early next week.

**Comment by Slach** [comment]
ok. i hope my PR will be merged ;)

File: .gitignore [hunk 5]
```diff
@@ -13,3 +14,7 @@ vagrant/vagrant-ssh-key.pub
 Godeps/_workspace
 .gopath/
 main
+.idea/
+*.deb
+*.pcap
+*.log
```
Comment (shlomi-noach) [inline .gitignore:20]: :+1:

File: Vagrantfile [hunk 6]
```diff
@@ -16,8 +16,8 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
   config.vm.box = BOX
   config.vm.box_download_insecure = true
   config.vm.box_check_update = false
-  config.vm.synced_folder '.', '/orchestrator', type: 'rsync',
-    rsync__auto: true
+  config.vm.synced_folder '.', '/orchestrator'
+  #, type: 'rsync', rsync__auto: true
```
Comment (shlomi-noach) [inline Vagrantfile]: Can you please explain this change?

File: build.sh [hunk 7]
```diff
@@ -87,7 +87,7 @@ function oinstall() {
   cd  $mydir
   gofmt -s -w  go/
   rsync -qa ./resources $builddir/orchestrator${prefix}/orchestrator/
-  rsync -qa ./conf/orchestrator-sample.* $builddir/orchestrator${prefix}/orchestrator/
+  rsync -qa ./conf/orchestrator-sample*.conf.json $builddir/orchestrator${prefix}/orchestrator/
```
Comment (shlomi-noach) [inline build.sh:90]: :+1:

File: build.sh [hunk 8]
```diff
@@ -134,12 +134,14 @@ function package() {
   esac
 
   echo "---"
-  if cat /etc/centos-release | grep 'CentOS release 6' ; then
-    rm ${TOPDIR:-?}/orchestrator*.deb
-    rm ${TOPDIR:-?}/orchestrator*.tar.gz
-    # n CentOD 6 box: we only want the rpms for CentOS6
-    # Add "-centos6" to the file name.
-    ls ${TOPDIR:-?}/*.rpm | while read f; do centos_file=$(echo $f | sed -r -e "s/^(.*)-${RELEASE_VERSION}(.*)/\1-centos6-${RELEASE_VERSION}\2/g") ; mv $f $centos_file ; done
+  if [[ -e /etc/centos-release ]]; then
```
Comment (shlomi-noach) [inline build.sh]: better: `-f`

File: conf/orchestrator-sample-sqlite.conf.json [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  "DetachLostSlavesAfterMasterFailover": true,
+  "ApplyMySQLPromotionAfterMasterFailover": false,
+  "MasterFailoverDetachSlaveMasterHost": false,
+  "MasterFailoverLostInstancesDowntimeMinutes": 0,
+  "PostponeSlaveRecoveryOnLagMinutes": 0,
+  "OSCIgnoreHostnameFilters": [],
+  "GraphiteAddr": "",
+  "GraphitePath": "",
+  "GraphiteConvertHostnameDotsToUnderscores": true
+}
```
Comment (shlomi-noach) [inline conf/orchestrator-sample-sqlite.conf.json:120]: :+1:

File: go/agent/agent_dao.go [hunk 10]
```diff
@@ -693,7 +693,7 @@ func executeSeed(seedId int64, targetHostname string, sourceHostname string) err
 	}
 
 	seedFromLogicalVolume := sourceAgent.LogicalVolumes[0]
-	seedStateId, _ = submitSeedStateEntry(seedId, fmt.Sprintf("Mounting logical volume: %s", seedFromLogicalVolume.Path), "")
+	seedStateId, _ = submitSeedStateEntry(seedId, fmt.Sprintf("%s Mounting logical volume: %s", sourceHostname, seedFromLogicalVolume.Path), "")
```
Comment (shlomi-noach) [inline go/agent/agent_dao.go:696]: :+1:

File: go/agent/agent_dao.go [hunk 11]
```diff
@@ -722,8 +722,8 @@ func executeSeed(seedId int64, targetHostname string, sourceHostname string) err
 	seedStateId, _ = submitSeedStateEntry(seedId, fmt.Sprintf("%s will now receive data in background", targetHostname), "")
 	ReceiveMySQLSeedData(targetHostname, seedId)
 
-	seedStateId, _ = submitSeedStateEntry(seedId, fmt.Sprintf("Waiting some time for %s to start listening for incoming data", targetHostname), "")
-	time.Sleep(2 * time.Second)
+	seedStateId, _ = submitSeedStateEntry(seedId, fmt.Sprintf("Waiting %d seconds for %s to start listening for incoming data", config.Config.SeedWaitSecondsBeforeSend, targetHostname), "")
+	time.Sleep(time.Duration(config.Config.SeedWaitSecondsBeforeSend) * time.Second)
```
Comment (shlomi-noach) [inline go/agent/agent_dao.go:726]: :+1:

File: go/db/generate_base.go [hunk 12]
```diff
@@ -138,7 +138,7 @@ var generateSQLBase = []string{
 		  last_checked timestamp NULL DEFAULT NULL,
 		  last_seen timestamp NULL DEFAULT NULL,
 		  mysql_port smallint(5) unsigned DEFAULT NULL,
-		  count_mysql_snapshots smallint(5) unsigned NOT NULL,
+		  count_mysql_snapshots smallint(5) unsigned NOT NULL DEFAULT 0,
```
Comment (shlomi-noach) [inline go/db/generate_base.go]: 🚫 Must change: existing SQL must never ever change. The most we can do is add a modification in `generate_patches.go`. However, modifying a column is not supported by `sqlite` so this change must unfortunately go away.

The value is implicitly `0` anyway, what is the concern?

**Review by shlomi-noach — CHANGES_REQUESTED** [review]
Thank you for the PR! and I'm sorry for the slow response; I have had some busy days and will have yet more busy days ahead of me.
One critical change, and some questions.

File: Vagrantfile (hunk 6 above)
Comment (Slach) [reply Vagrantfile]: rsync__auto want work on my Vagrant environment ;(
when i change some sources on host OS, it's not rsynced into guest maching
ok. i rollback this change

Comment (shlomi-noach) [reply Vagrantfile]: Sorry, please elaborate. Did you introduce `rsync__auto` in the first place?

File: go/db/generate_base.go (hunk 12 above)
Comment (Slach) [reply go/db/generate_base.go]: oh. ok. sorry i will rollback this change

Comment (Slach) [reply go/db/generate_base.go]: yes concern is 
count_mysql_snapshots not used when try insert into ```host_agent``` table
and sql query is failed when agent submitting

File: Vagrantfile (hunk 6 above)
Comment (Slach) [reply Vagrantfile]: i try use following variants
1)
```
 config.vm.synced_folder '.', '/orchestrator', type: 'rsync'
```

it's require run ```vagrant reload``` when i change any file on my host machine
2)
```
 config.vm.synced_folder '.', '/orchestrator', type: 'rsync',  rsync__auto: true
```
it not worked, and i don't investigate why
then i just commented this options and default sharing mode over vboxfs works as i expected

File: build.sh [hunk 13]
```diff
@@ -134,12 +134,14 @@ function package() {
...
-    # Add "-centos6" to the file name.
-    ls ${TOPDIR:-?}/*.rpm | while read f; do centos_file=$(echo $f | sed -r -e "s/^(.*)-${RELEASE_VERSION}(.*)/\1-centos6-${RELEASE_VERSION}\2/g") ; mv $f $centos_file ; done
+  if [[ -f /etc/centos-release ]]; then
+      if cat /etc/centos-release | grep 'CentOS release 6' ; then
+        rm ${TOPDIR:-?}/orchestrator*.deb
+        rm ${TOPDIR:-?}/orchestrator*.tar.gz
+        # n CentOD 6 box: we only want the rpms for CentOS6
+        # Add "-centos6" to the file name.
+        ls ${TOPDIR:-?}/*.rpm | while read f; do centos_file=$(echo $f | sed -r -e "s/^(.*)-${RELEASE_VERSION}(.*)/\1-centos6-${RELEASE_VERSION}\2/g") ; mv $f $centos_file ; done
+      fi
```
Comment (shlomi-noach) [inline build.sh:144]: :+1:

File: vagrant/admin-build.sh [hunk 14]
```diff
@@ -1,23 +1,34 @@
...
-  cp /usr/local/orchestrator/orchestrator-sample.conf.json /etc/orchestrator.conf.json
   /usr/sbin/service orchestrator start
 
 fi
 
 echo '* * * * * root /usr/bin/orchestrator -c discover -i db1' > /etc/cron.d/orchestrator-discovery
 
 # Discover instances
-/usr/bin/orchestrator -c discover -i localhost
+/usr/bin/orchestrator --verbose --debug --stack -c redeploy-internal-db
```
Comment (shlomi-noach) [inline vagrant/admin-build.sh]: What happened that the above became necessary? Current logic should make it safe to run `orchestrator` without specifying `redeploy-internal-db`.

Comment (Slach) [reply vagrant/admin-build.sh]: ohh, sorry, i multiple run ```vagrant up``` and add redeploy-internal-db to try fix errors when using sqlite backend
i revert this changes right now

**Comment by Slach** [comment]
@shlomi-noach would this PR ready for merge? i think i resolve any recomendation

**Review by shlomi-noach — APPROVED** [review]

**Comment by Slach** [comment]
thanks!

**Comment by shlomi-noach** [comment]
Thank you!

**Pull request merged**